Added animations for some electromagnetic phenomena

SunTzunami opened the pull request:
I made some basic animations using FURY that render some electromagnetic phenomena like-

1. Propagation of an an electromagnetic wave.

![em_wave](https://user-images.githubusercontent.com/65067354/107858093-5eb48b80-6e58-11eb-925d-d8a17dd8a9a7.gif)

2. Helical motion of a charged particle under the influence of a combined magnetic and electric field.

![helical_motion](https://user-images.githubusercontent.com/65067354/107858126-902d5700-6e58-11eb-9297-518f1268157b.gif)

Nibba2018 commented on docs/examples/Electromagnetic Phenomena/emwave_animation.py:
```diff
@@ -0,0 +1,169 @@
+phase with the electric field, but in the ±y-direction.
+
+Function of the sinusoid used in the animation = sin(k*x - w*t + d)
+Where, k:wavenumber, x:abscissa, w:angular frequency, t:time, d:phase angle
+
+"""
+
+###############################################################################
+# Importing necessary modules
```
Merge the comment block with the multi-line string block.

Nibba2018 commented on docs/examples/Electromagnetic Phenomena/emwave_animation.py:
```diff
@@ -0,0 +1,169 @@
+w = 0.1
+d = 0.002
+
+###############################################################################
+# Creating a scene object and configuring the camera's position
+scene = window.Scene()
+scene.set_camera(position=(6, 5, 10), focal_point=(0.0, 0.0, 0.0),
```
add a newline after the comment block ends.

Nibba2018 commented on docs/examples/Electromagnetic Phenomena/emwave_animation.py:
```diff
@@ -0,0 +1,169 @@
+                           order_transparent=True)
+showm.initialize()
+
+###############################################################################
+# Creating a yellow colored arrow to show the direction of propagation of
+# electromagnetic wave
+centers = np.array([[-3, 0, 0]])
+directions = np.array([[1, 0, 0]])
+heights = np.array([6.4])
```
same as above and the rest below them.

Nibba2018 commented on docs/examples/Electromagnetic Phenomena/emwave_animation.py:
```diff
@@ -0,0 +1,169 @@
+# using the update_coordinates function. The wave is rendered here.
+
+
+def timer_callback(_obj, _event):
+    global pts, pts2, colors, rate, t, dt, w, d, k
+    t += dt
+
```
`rate` is unused.

SunTzunami replied:
Removed rate. Thanks for pointing that out. It remained from an earlier version of the code.

Nibba2018 commented on docs/examples/Electromagnetic Phenomena/helical_motion.py:
```diff
@@ -0,0 +1,152 @@
+path.
+
+In this animation, there's a magnetic and an electric field present in +x
+direction under whose influence the charged particle follows a helical path.
+
+"""
+
+###############################################################################
+# Importing necessary modules
```
Merge the two blocks.

Nibba2018 commented on docs/examples/Electromagnetic Phenomena/helical_motion.py:
```diff
@@ -0,0 +1,152 @@
+d = 0.002
+
+
+###############################################################################
+# Creating a scene object and configuring the camera's position
+scene = window.Scene()
+scene.zoom(1.2)
+scene.set_camera(position=(10, 12.5, 19), focal_point=(3.0, 0.0, 0.0),
```
Add a newline after comment block.

Nibba2018 commented on docs/examples/Electromagnetic Phenomena/helical_motion.py:
```diff
@@ -0,0 +1,152 @@
+                           order_transparent=True)
+showm.initialize()
+
+
+###############################################################################
+# Creating a blue colored arrow to show the direction of propagation of
+# electromagnetic wave
+color_arrow = window.colors.blue  # color of the arrow can be manipulated
+centers = np.array([[0, 0, 0]])
```
same as above and below.
I think it should be `charged particle` instead of `EM wave`.

SunTzunami replied:
Thanks for pointing this out. It's supposed to represent the direction of magnetic and electric field. I rectified it.

Nibba2018 reviewed (changes requested):
Hey @SunTzunami , thanks for the PR. I liked your examples and would recommend you to rename the folder as `electromagnetic_phenomena` instead of the current case. Also prepend your file names, both python scripts and window.record png files with `viz_` in order to help index the generated doc(scripts and png files should have the same name).

Apart from that, according to the right hand thumb rule, I think the direction of your charged particle should be `anti-clockwise` instead of it being `clockwise` (correct me if I am wrong). 

Also I have mentioned some minor changes to fix the formatting below.

skoudoro commented on docs/examples/Electromagnetic Phenomena/emwave_animation.py:
```diff
@@ -0,0 +1,169 @@
+# changing with time
+
+
+def update_coordinates(k, w, t, d):
```
In general, you should avoid to have 1 letter as a variable name. Even if it is really nice that you describe them above. Can you update the whole file with the correct name. Thank you!

SunTzunami replied:
I renamed the variables. Thank you!

skoudoro commented on docs/examples/Electromagnetic Phenomena/helical_motion.py:
```diff
@@ -0,0 +1,152 @@
+
+radius_particle = 0.08
+radius_path = 0.04
+u = 0.09
+a = 0.004
+t = 0
+dt = 0.09
+w = 0.1
+d = 0.002
```
same comments as above

skoudoro commented on docs/examples/Electromagnetic Phenomena/emwave_animation.py:
```diff
@@ -0,0 +1,169 @@
+
+###############################################################################
+# end is used to decide when to end the animation
+end = 1000
```
For doc generation, can you make it shorter. thanks!

SunTzunami replied:
I suppose you meant that I reduce the value of end? I reduced end from 1000 to 300.

skoudoro replied:
Yes, and thank you

skoudoro reviewed (commented):
Hi @SunTzunami,

Really Nice tutorial, Good Job!  

Can you remove this subfolder (`Electromagnetic Phenomena`). Just let your demo in the examples folder for now. We will reshape the whole documentation later.

Also, it would be good to have those demos with the line actor instead of the point actor.
Can you try it?

pep8speaks commented:
Hello @SunTzunami! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`docs/examples/viz_emwave_animation.py`](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py):

> [Line 41:1](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L41): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 41:3](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L41): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 41:7](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L41): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 41:13](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L41): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 49:3](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L49): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 49:5](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L49): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 49:7](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L49): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 56:7](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L56): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 56:49](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L56): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 56:80](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_emwave_animation.py#L56): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 79 characters)

* In the file [`docs/examples/viz_helical_motion.py`](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py):

> [Line 29:3](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L29): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 29:7](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L29): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 29:13](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L29): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 38:3](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L38): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 38:5](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L38): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 38:7](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L38): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 50:7](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L50): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 50:49](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L50): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 50:80](https://github.com/fury-gl/fury/blob/c9dfddc8bf2c80b8b30416d776e86993577f0c5f/docs/examples/viz_helical_motion.py#L50): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)

SunTzunami commented:
Hi @Nibba2018, @skoudoro! I really appreciate the feedback. Thank you! 
@Nibba2018, I made the changes to the comments as you instructed and renamed the files too. The path of the particle will be clockwise (I think the confusion arose because I forgot to mention that it's positively charged).  I've attached a screenshot of a physics textbook which has a similar concept.
![helical_path](https://user-images.githubusercontent.com/65067354/108164919-b2a0c800-7117-11eb-9b1d-a2855d640e6f.PNG)

@skoudoro, I renamed the variables and made their names more descriptive. I tried using line actor for the animation but could only implement it for the helical motion animation.

SunTzunami commented:
I don't know why flake8 errors are arising, I didn't have them in my pc before I pushed the files.

SunTzunami commented on docs/examples/viz_emwave_animation.py:
```diff
@@ -0,0 +1,184 @@
+#          the radius_waves small but kindly note that higher values for
+#          npoints will slow down the rendering process
+#          (default = 800)
+<<<<<<< HEAD:docs/examples/viz_emwave_animation.py
```
These lines weren't there when I pushed, should I remove them? Or are they normal? @Nibba2018 @skoudoro ?

SunTzunami commented on docs/examples/viz_emwave_animation.py:
```diff
@@ -0,0 +1,184 @@
+# dt: value by which t is incremented for each call of timer_callback
+# w: angular frequency (default = 0.1)
+# d: phase angle (default = 0.002)
+>>>>>>> 7a572769240e56179dd34f087fa1bccfd3bf0e33:docs/examples/Electromagnetic Phenomena/emwave_animation.py
```
This too

SunTzunami commented on docs/examples/viz_emwave_animation.py:
```diff
@@ -0,0 +1,184 @@
+#             timer_callback (default = 0.1)
+# angular_frq: angular frequency (default = 0.1)
+# phase_angle: phase angle (default = 0.002)
+=======
```
This too.

Nibba2018 commented:
@SunTzunami Yeah my bad, clockwise direction is correct.

skoudoro commented:
Thank you for the update.

Could you just explain what was your issue with `line actor` when you tried to implement it in `viz_emwave_animation.py`?

skoudoro commented:
Also, Could you create a PR with your gif files in this repo: https://github.com/fury-gl/fury-communication-assets?

Thank you @SunTzunami

SunTzunami commented:
> Thank you for the update.
> 
> Could you just explain what was your issue with `line actor` when you tried to implement it in `viz_emwave_animation.py`?

Hi @skoudoro ! Thanks for the review! I didn't know how to update coordinates for line actor I then spent quite some time figuring out how to update coordinates for line actor. It works now.  :) 
The idea of using line actor was nice @skoudoro. It looks very nice now. Thank you!
Also, there was an error in the EM wave animation (the direction of the arrow was opposite to what it should be). I rectified it.
I'll create a PR with the gifs (with new animations that use line actor).

skoudoro reviewed (approved):
Great, Thank you for the update and for using the line actor.

+1 for merging.  Is it good for you too @Nibba2018?

Nibba2018 commented:
@SunTzunami  @skoudoro Looks good to me too!

Nibba2018 commented:
![em](https://user-images.githubusercontent.com/29832615/108412609-ff87ba00-724f-11eb-9027-b24d5b287157.gif)
Isn't the direction of the arrow in the opposite direction?

Nibba2018 reviewed (changes requested):
> Also, there was an error in the EM wave animation (the direction of the arrow was opposite to what it should be). I rectified it.

 i guess you haven't pushed these changes.

SunTzunami commented:
> > Also, there was an error in the EM wave animation (the direction of the arrow was opposite to what it should be). I rectified it.
> 
> i guess you haven't pushed these changes. @SunTzunami
 @Nibba2018 [I did it](https://github.com/fury-gl/fury/pull/376/commits/839c323f6dbd82344b67b1ced93347ec4262069f). I changed the camera's location too.

Earlier-
![em_wave](https://user-images.githubusercontent.com/65067354/108414392-2e9f2b00-7252-11eb-8cd7-a6f847480289.gif)

Now-
![em_wave](https://user-images.githubusercontent.com/65067354/108414474-48407280-7252-11eb-9d88-be5400424d89.gif)
